### PR TITLE
Add option for additional resource attributes with regex to trace exporter

### PR DESCRIPTION
**Pull request by dhpollack**
The [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#specifying-resource-information-via-an-environment-variable) allows one to specify additional attributes with the special environmental variable `OTEL_RESOURCE_ATTRIBUTES`.  This includes those attributes into the span exported to GCP.

The goal of this PR is to easy add "global" attributes to all spans in a given service so that they can be filtered easily in GCP.  But also it is not intuitive that you can add these attributes to spans, but they don't get exported with the `CloudTraceSpanExporter`.

Signed-off-by: David Pollack <david@da3.net>

### CODE REVIEW

**Comment by aabmass** [comment]
Can you remove the `OTEL_RESOURCE_ATTRIBUTES` parts since that is already covered by OTel python? We are thinking the other part of this PR would be great to add, with some configuration. Can you update the `CloudTraceSpanExporter` constructor to accept a regex for the resource attributes to pass through. If the option isn't given, it shouldn't change the existing behavior

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (aabmass) [inline opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:384]: Only certain resource attributes are recognized by Google with this `g.co/r/` format. The ones that pass the regex can just be added straight in without changing the keys.

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (aabmass) [inline opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: This isn't needed, this is done automatically by OTel. The resource this exporter sees below will already have the fields from the environrment variable.

I was thinking we could just add a regex here, where resource keys that match the regex will be passed on as span attributes.

**Review by aabmass — CHANGES_REQUESTED** [review]

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: But we don't know what the resource keys are because they are based on the environmental variable.  So you want only a few pre-set list of resource attribute to be passed on to the span attribute?  Doesn't that kind of defeat the purpose of allowing people to use the `OTEL_RESOURCE_ATTRIBUTES` environmental variable?

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:384]: This was extracted [directly from `main`](https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/blob/master/opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py#L370).  I can change this, but what format at you looking for exactly.

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: I could just extract the keys here, but I thought it made a little more sense for all the inputs to `_extract_resources` to be resources instead of a hybrid between `Resource` and `Attribute` / `Dict[str, str]`.  The other idea is to add it to `resources_and_attrs` in line 166 directly with `**self.otel_env_resource`, but then testing would require a bit of refactoring.

Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: In my use case, I gave some examples to simplify what I actually wanted.  The truth is we use a non-standard name called `app_name` and I would probably use that instead of `service.name`.  But I didn't want this PR to be able my specific case, but more about generally adding extra attributes to the spans.  The idea being that one could add arbitrary attributes (even ridiculously named non-standard attributes like `app_name`) that would show up in the GCP interface as labels which can be used for filtering by using the environmental variable `OTEL_RESOURCE_ATTRIBUTES`.

Apologies if I wasn't clear on this earlier.

Comment (aabmass) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: No worries. For the general use case, I think we can address it again later if someone can't solve it with a more restrictive regex.

Would either a filter function that returns true/false if the resource key should be included, or a regex pattern for resource keys to include solve your specific issue?

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (aabmass) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:384]: The `g.co/r/` attributes should only correspond to monitored resource labels (e.g. [gce_instance](https://cloud.google.com/monitoring/api/resources#tag_gce_instance)). The additional resource labels we are putting can just be copied over from the OTel resource in the span.

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:107]: Ok, I get it.  Yea, that would work for me.  I'll implement the regex.

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py]: @aabmass let me know if this is what you want now.  For `gce_instance` I kept it the same as before but for `gke_container` I only use the `ot_resource_key` (for example `k8s.cluster.name`).

**Comment by dhpollack** [comment]
@aabmass I think this is ready now

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (aabmass) [inline opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:93]:
```suggestion
        resource_regex: Resource attributes with keys matching this regex will be
          added to exported spans as labels (default: None).
```

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (aabmass) [inline opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py:391]: I don't think this if/else is needed, we still special case both keys of `OT_RESOURCE_ATTRIBUTE_TO_GCP`.

File: opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py
Comment (aabmass) [inline opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py:626]: I notice we are using `regex.match()` above, in which case the `.*` isn't required here. What do you think is more useful/intuitive, `fullmatch()` or `match()`?

**Review by aabmass — COMMENTED** [review]
Nice, looks good! Just add back the gke_cluster case and we should be good

File: opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py:626]: I personally don't use python regexes that often, so I wouldn't read too much into my intention based on this regex (i.e. this is probably an inefficient / "bad" regex because I don't write them that often).  As a general rule, I would use which ever one seems more prevalent.  But I would certainly defer to you as to what you prefer on this one.

File: opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py
Comment (dhpollack) [reply opentelemetry-exporter-gcp-trace/src/opentelemetry/exporter/cloud_trace/__init__.py]: ok, I misread you previous comment.  changed this back.

File: opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py
Comment (aabmass) [reply opentelemetry-exporter-gcp-trace/tests/test_cloud_trace_exporter.py:626]: I think the same sort of thing could be accomplished with all of them. Lets leave it as match for now 👍

**Review by aabmass — APPROVED** [review]
Thanks for working through my comments, LGTM!

**Comment by aabmass** [comment]
@dhpollack could you rename this PR to something more descriptive of what we landed on, and add an entry to the changelog file? I'll merge after that 😃

**Comment by dhpollack** [comment]
> @dhpollack could you rename this PR to something more descriptive of what we landed on, and add an entry to the changelog file? I'll merge after that smiley

@aabmass done!  Thanks for all the help.

**Pull request merged**
